model/src/app_data.rs

```diff
@@ -0,0 +1,135 @@
+        // The ipfs cid can be calculated by
+        // const RAW: u64 = 0x55;
+        // let hash = Code::Sha2_256.digest(&string.into_bytes());
+        // let cid = Cid::new_v1(RAW, hash);
+        // In order to avoid json duplication, we are deriving the hash from the json object
+        let hash = Code::Sha2_256.digest(serde_json::ser::to_string(&self.0.clone())?.as_bytes());
+        let array: [u8; 32] = hash
+            .digest()
+            .try_into()
+            .expect("sha256 hash unexpected length");
+        Ok(H256::from(array))
+    }
+    pub fn get_app_data(&self) -> Result<AppData, serde_json::Error> {
+        serde_json::from_value(self.0.clone())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use serde_json::json;
+    #[test]
+    fn deserialization_and_back() {
+        let value = json!(
+        {
+            "appCode": "CowSwap",
+            "version": "1.0.0",
+            "metadata": {
+              "referrer": {
+                "address":  "0x424a46612794dbb8000194937834250dc723ffa5",
+                "version": "0.3.4",
+              }
+            }
+        }
+        );
+        let expected = AppData {
+            version: String::from("1.0.0"),
+            app_code: Some(String::from("CowSwap")),
+            metadata: Some(Metadata {
+                referrer: Some(Referrer {
+                    address: "0x424a46612794dbb8000194937834250dc723ffa5"
+                        .parse()
+                        .unwrap(),
+                    version: String::from("0.3.4"),
+                }),
+            }),
+        };
+        let deserialized = AppDataBlob(value.clone());
+        assert_eq!(deserialized.get_app_data().unwrap(), expected);
+        let serialized = serde_json::to_value(expected).unwrap();
+        assert_eq!(serialized, value);
+    }
+
+    #[test]
+    fn deserialization_and_back_for_nearly_empty_data() {
+        let value = json!(
+        {
+            "appCode": null,
+            "version": "0.1",
+            "metadata": null
+        }
+        );
+        let expected = AppData {
+            app_code: None,
+            version: String::from("0.1"),
+            metadata: None,
+        };
+        let deserialized = AppDataBlob(value.clone());
+        assert_eq!(deserialized.get_app_data().unwrap(), expected);
+        let serialized = serde_json::to_value(expected).unwrap();
+        assert_eq!(serialized, value);
+    }
+    #[test]
+    fn test_hash_calculation() {
+        let json = json!(
+        {
+            "appCode": "CowSwap",
+            "version": "1.0.0",
+            "metadata": {
+              "referrer": {
+                "address":  "0x424a46612794dbb8000194937834250dc723ffa5",
+                "version": "0.3.4",
+              }
+            }
+        }
+        );
+        let app_data_blob = AppDataBlob(json);
+        let expected: H256 = "0xd0b33abbeb0349c0b9d8d1a9e1c38b28efbc9c48842c26697a41a463d019cb16"
```

Comment:
Shouldn't this be:

`0x474352A1F08C1FD73B7A450B8EC0B04DC164B2C131717AF15E71AB578B5C3A4F`

https://gateway.pinata.cloud/ipfs/QmT8qCMvCtXtWpMeNwk7AeHHUnJNfv83DxDawLMEpMz694

See: 
https://cid.ipfs.io/#QmT8qCMvCtXtWpMeNwk7AeHHUnJNfv83DxDawLMEpMz694